### ISSUE TITLE
suggestion: make alyx instance into an object, replicate "dat" methods in it

### ISSUE THREAD

**Issue by nsteinme** [cortex-lab/alyx-matlab#30]
So whereas before we'd do: 
```
>> dat.listSubjects
>> dat.expFilePath(subjects, date, number, type)
>> dat.newExp(subject)
```
etc, we now would have:
```
>> ai = alyx.loginWindow(); % or other constructor
>> ai.listSubjects % can list the living subjects from alyx since it has credentials already
>> ai.expFilePath(subject, date, number, type) % can find these for you
```
etc. Thoughts? Then we could phase out "dat" altogether. I don't think it would affect existing usage of alyx instances that are currently structs, and would allow a security improvement (protecting the token).

**Comment by kdharris101** [comment]
From my limited understanding of what dat does, this sounds like a good idea.

**Comment by nippoo** [comment]
This is definitely something I had been planning to do for a while - my idea was to make these pretty lightweight functions and implement most of the logic in REST.

**Comment by k1o0** [comment]
I may make a start on this next week then...

**Comment by nsteinme** [comment]
Thanks @k1o0 .

@rossant Can you comment on @nippoo 's suggestion to implement most logic in REST? The main one in question is "expFilePath" which should return the repository_path and relative_path of the requested dataset. My memory is that we found that this was faster in SQL than with REST API queries (hence [this implementation](https://github.com/cortex-lab/alyx-matlab/blob/master/%2Balyx/expFilePath.m)). Can you check whether that is still true and see about whether it could be quickly implemented server-side somehow instead? It will indeed presumably be more robust to breaking if it's server-side rather than an sql query.

**Comment by rossant** [comment]
@nsteinme  yes it would be better to implement the logic server-side with an adequate endpoint. Which endpoint exactly would that be, for example for expFilePath? I see no reason why it shouldn't be as fast with REST -- at worse we can always put the SQL query server-side in Python behind an endpoint.

**Comment by nsteinme** [comment]
I guess it should be a new endpoint - let me make an issue for it and
describe the intended behavior.

**Comment by k1o0** [comment]
Alyx is now a class on the alyx-as-class branch.  Still needs integrating with Rigbox though.

**Comment by k1o0** [comment]
This is done now, and almost fully documented (@nsteinme could you document expTimelinePlot.m and getExpMeta.m?).   The training rigs will now be operating on the alyx-as-class branch, so far everything appears to be working.  Now we just need to sort out the relevant endpoints for integrating some of the dat functions.

**Comment by k1o0** [comment]
I think we should revisit the question of what to do with the dat package.  At one time it was decided to keep paths and file functions entirely separate from the database.  I still think this is the simplest solution.  Otherwise the only hurdles to phasing out dat entirely are...
1. how to indicate which repo serves what purpose, i.e. which is the 'main' one.  When saving a file, /data-repository can return a list of locations but there's no way of determining which repo is to save to;
2. how to indicate where on the file server to save the data.  Currently Alyx can return zubjects as a repo, but not the full save path.

Other than those two points, it would be a fine solution to have the files saved according to Alyx (with the 'LocalExpData' path used as a fallback).  The advantage to having Alyx determine the save location is that you would not have to update the GitHub code and pull on every machine.

**Comment by k1o0** [comment]
We just discussed this in the data architecture working group and it seems this is something we won't be able to implement.  Currently the database can only tell you the hostname where a particular dataset is stored.  It can't tell you where the data are or how to access the repository. @oliche considers the precise location of files on zserver/zubjects (and their access protocols) outside the scope of alyx.  As we currently use user defined paths to both save and access data in cortexlab, and alyx is at this moment specifically engineered for the IBL data pipeline, we will simply continue to use local paths for most of our data organization.

@nsteinme I think I failed to express your thoughts on this cogently, so feel free to reiterate them.  I will hack a solution to pull request #56 using regular expressions but there won't be any more integration of alyx for data organization.
